docs/external.rst

```diff
@@ -1,68 +1,140 @@
-  m = 7
 
-  # Taichi fields
-  val = ti.field(ti.i32, shape=(n, m))
-  vec = ti.Vector.field(3, dtype=ti.i32, shape=(n, m))
-  mat = ti.Matrix.field(3, 4, dtype=ti.i32, shape=(n, m))
+**Import data from NumPy array to Taichi fields** via ``from_numpy()``.
+This allow people to initialize Taichi fields via numpy array, e.g.:
+
+.. code-block:: python
+
+   x = ti.field(ti.f32, 4)
+   x_np = np.array([1, 7, 3, 5])
+   x.from_numpy(x_np)
+   print(x[0])  # 1
+   print(x[1])  # 7
+   print(x[2])  # 3
+   print(x[3])  # 5
+
+
+API reference
+-------------
+
+We provide interface to share data with both NumPy and PyTorch.
+
+Interacting with NumPy
+**********************
+
+.. function:: field.to_numpy()
+
+   :parameter field: (ti.field, ti.Vector.field or ti.Matrix.field) The field
+
+   :return: (np.array) The numpy array containing the currrent data in ``x``.
+
+.. function:: field.from_numpy(array)
+
+   :parameter field: (ti.field, ti.Vector.field or ti.Matrix.field) The field
+
+   :parameter array: (np.array) The numpy array containing data to initialize the field
 
-  # Scalar
-  arr = np.ones(shape=(n, m), dtype=np.int32)
 
-  val.from_numpy(arr)
+Interacting with PyTorch
+************************
 
-  arr = val.to_numpy()
+.. function:: field.to_torch(device = None)
 
-  # Vector
-  arr = np.ones(shape=(n, m, 3), dtype=np.int32)
+   :parameter field: (ti.field, ti.Vector.field or ti.Matrix.field) The field
+   :parameter device: (torch.device) the device where torch tensor is created
 
-  vec.from_numpy(arr)
+   :return: (torch.Tensor) The torch tensor containing the currrent data in ``x``.
 
-  arr = np.ones(shape=(n, m, 3, 1), dtype=np.int32)
-  vec.from_numpy(arr)
+.. function:: field.from_torch(tensor)
 
-  arr = vec.to_numpy()
-  assert arr.shape == (n, m, 3)
+   :parameter field: (ti.field, ti.Vector.field or ti.Matrix.field) The field
 
-  arr = vec.to_numpy(keep_dims=True)
-  assert arr.shape == (n, m, 3, 1)
+   :parameter tensor: (torch.Tensor) The torch tensor containing data to initialize the field
 
-  # Matrix
-  arr = np.ones(shape=(n, m, 3, 4), dtype=np.int32)
 
-  mat.from_numpy(arr)
+The shape of external array to interact with Taichi fields
+----------------------------------------------------------
 
-  arr = mat.to_numpy()
-  assert arr.shape == (n, m, 3, 4)
+Taichi fields have shapes (see :ref:`scalar_tensor`). NumPy array also have shapes.
+So what's the relationship between them when using the ``to_numpy()`` and ``from_numpy()``?
 
+- For scalar fields, **the shape of NumPy array should be exactly the same with the Taichi field**:
 
-TODO: add API reference
+.. code-block:: python
+
+   field = ti.field(ti.i32, shape=(233, 666))
+   field.shape  # (233, 666)
+
+   array = field.to_numpy()
+   array.shape  # (233, 666)
+
+   field.from_numpy(array)  # the input array must be of shape (233, 666)
+
+
+- For vector fields, if the vector is ``n``-D, then **the shape of NumPy array should be** ``(*field_shape, vector_n)``:
```

Comment:
```suggestion
- For vector fields, if the vector is ``n``-D, then **the shape of NumPy array is** ``field_shape + (vector_n,)``:
```

Reply:
It does work:
```py
>>> shape = (2, 3)
>>> (*shape, 4)
(2, 3, 4)
```
I would use the original one since it's more clear at first glance.

Reply:
Yeah I know that star expression works fine here, but I'm not sure how many people can easily understand this fancy usage.

Reply:
My idea was people may confuse tuple `+` with scalar `+`..

Reply:
Both `tuple + scalar` and `*tuple` are confusion. What about a sentence of explanation of astroid like `, where the asteroid symbol before the tuple means xxx`

Reply:
I agree. Maybe we should simply use examples and text description instead. No need to test users' Python skills here.

Reply:
Yeah we have examples, see L117-124.

Reply:
OK sounds good.